pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,350 @@
+}
+
+func (b *lockableBackend) Update(ctx context.Context, stackRef backend.StackReference,
+	op backend.UpdateOperation) (engine.ResourceChanges, result.Result) {
+	err := b.Lock(stackRef)
+	if err != nil {
+		return nil, result.FromError(err)
+	}
+	defer b.Unlock(stackRef)
+	return b.lb.Update(ctx, stackRef, op)
+}
+
+func (b *lockableBackend) UpdateStackTags(ctx context.Context,
+	stackRef backend.StackReference, tags map[apitype.StackTagName]string) error {
+	err := b.Lock(stackRef)
+	if err != nil {
+		return err
+	}
+	defer b.Unlock(stackRef)
+	return b.lb.UpdateStackTags(ctx, stackRef, tags)
+}
+
+func (b *lockableBackend) checkForLock(ctxt context.Context, stackRef backend.StackReference) error {
+	allFiles, err := listBucket(b.lb.bucket, b.lockDir())
+	if err != nil {
+		return err
+	}
+
+	var lockKeys []string
+	for i := len(allFiles) - 1; i >= 0; i-- {
+		if allFiles[i].IsDir {
+			continue
+		}
+		fileKey := allFiles[i].Key
+		if b.isLockForThisStack(fileKey, stackRef) {
+			lockKeys = append(lockKeys, fileKey)
+		}
+	}
+
+	if len(lockKeys) > 0 {
+		errorString := fmt.Sprintf("the stack is current locked by %v lock(s). Either wait for the other "+
+			"process(es) to end or manually delete the lock file(s).", len(lockKeys))
+
+		for _, lock := range lockKeys {
+			content, err := b.lb.bucket.ReadAll(ctxt, lock)
+			if err != nil {
+				return err
+			}
+			l := &lockContent{}
+			err = json.Unmarshal(content, &l)
+			if err != nil {
+				return err
+			}
+
+			// this is kind of weird but necessary because url is a string and not a url.URL
+			url := b.lb.url + filepath.Join("/", lock)
+			errorString = errorString + fmt.Sprintf("\n  %v: created by %v", url, l.String())
+		}
+
+		return fmt.Errorf(errorString)
+	}
+	return nil
+}
+
+func (b *lockableBackend) checkForLockRace(stackRef backend.StackReference) error {
+	// Check the locks to make sure ONLY our lock exists. If we find multiple locks then we know
+	// we had a race condition and we should clean up and abort
+	allFiles, err := listBucket(b.lb.bucket, b.lockDir())
+	if err != nil {
+		return err
+	}
+
+	for i := len(allFiles) - 1; i >= 0; i-- {
+		file := allFiles[i].Key
+		if b.isLockForThisStack(file, stackRef) && file != b.lockPath(stackRef.Name()) {
+			return fmt.Errorf("another Pulumi execution grabbed the lock, please wait" +
+				"for the other Pulumi process to complete and try again")
+		}
+	}
+	return nil
+}
+
+func (b *lockableBackend) Lock(stackRef backend.StackReference) error {
+	ctxt := context.TODO()
+
+	err := b.checkForLock(ctxt, stackRef)
+	if err != nil {
+		return err
+	}
+
+	lockContent, err := newLockContent()
+	if err != nil {
+		return err
+	}
+
+	content, err := json.Marshal(lockContent)
+	if err != nil {
+		return err
+	}
+
+	err = b.lb.bucket.WriteAll(ctxt, b.lockPath(stackRef.Name()), content, nil)
+	if err != nil {
+		return err
+	}
+
+	err = b.checkForLockRace(stackRef)
+	if err != nil {
+		b.Unlock(stackRef)
```

Comment:
we should doc why we're unlocking here.  i think it's appropriate.  but it's necessary to understand the invariants here.    we might also need comments taht say if we add any more conditions here (and they fail) that we should unlock in those cases too.